src/inmanta/deploy/timers.py

```diff
@@ -0,0 +1,249 @@
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+import asyncio
+import logging
+import sys
+from collections.abc import Coroutine
+from typing import Any, Callable
+
+import inmanta.deploy.scheduler
+from inmanta.agent import config as agent_config
+from inmanta.data.model import ResourceIdStr
+from inmanta.deploy.state import BlockedStatus
+from inmanta.deploy.work import TaskPriority
+from inmanta.util import CronSchedule, ScheduledTask, Scheduler
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ResourceTimer:
+    def __init__(self, resource: ResourceIdStr, scheduler: "inmanta.deploy.scheduler.ResourceScheduler"):
+        """
+        :param resource: The resource for which to ensure periodic repairs.
+        """
+        self.resource: ResourceIdStr = resource
+        self.time_to_next_deploy = sys.maxsize
+
+        # "Inner" asyncio task responsible for queuing a Deploy task
+        self.trigger_deploy: asyncio.Task[None] | None = None
+        # "Outer" handle to the soon-to-be-called trigger_deploy
+        self.next_schedule_handle: asyncio.TimerHandle | None = None
+
+        self._resource_scheduler = scheduler
+
+    def ensure_timer(
+        self,
+        periodic_deploy_interval: int | None,
+        periodic_repair_interval: int | None,
+        is_dirty: bool,
+    ) -> None:
+        """
+        Make sure the underlying resource is marked for execution
+
+        :param periodic_deploy_interval: Per-resource deploy interval, assumed to be a positive int or None
+        :param periodic_repair_interval: Per-resource repair interval, assumed to be a positive int or None
+        :param is_dirty: Last known state of the resource at the timer install request time.
+        :param action_function: The function to execute
+        """
+
+        next_execute_time: int
+
+        if periodic_repair_interval is None:
+            periodic_repair_interval = sys.maxsize
+        if periodic_deploy_interval is None:
+            periodic_deploy_interval = sys.maxsize
+
+        action_function: Callable[[ResourceIdStr, int], Coroutine[Any, Any, None]]
+        if is_dirty:
+            next_execute_time = min(periodic_deploy_interval, periodic_repair_interval)
+            action_function = self._trigger_deploy_for_resource
+        else:
+            next_execute_time = periodic_repair_interval
+            action_function = self._trigger_repair_for_resource
+
+        if next_execute_time == sys.maxsize:
+            return
+
+        # A deploy is already scheduled for this resource
+        if self.next_schedule_handle is not None:
+            return
```

Comment:
I don't think this is correct, see comment  on comment



What about if we go from dirty to clean by receiving an event?

The deploy time hasn't fired, but it will and it shouldn't.

Reply:
Updated this to cancel and reschedule if the interval changed and to early return otherwise